File: diet_guard/_kuchnia_spread.py
"""Assign catering dishes to meal slots, in the provider's own order.

The panel returns each dish with a ``mealPriority`` (1..N: Śniadanie, II
śniadanie, Obiad, Podwieczorek, Kolacja) -- the order the catering intends them
to be eaten in.  That beats inferring an order from the dish names, and it
beats spreading them arithmetically, so this module sorts by priority and then
maps position onto the user's configured slots.

Counts rarely match: a 5-meal plan against the default 4 slots means two dishes
share the first slot.  The mapping is ``i * S // N``, which is monotonic
non-decreasing, keeps the first dish on the first slot and the last on the
last, and doubles up the earliest slots rather than dropping anything.

Integer ``//`` only, never ``round()``.  That is repo convention for slot
arithmetic (``docs/meal-schedule.md``): Python's banker's rounding and Dart's
half-away-from-zero disagree on ``.5``, and a slot one device offers while the
other does not is a checkpoint that can never be satisfied.  The parity risk is
theoretical *here* -- the phone has no importer and so no mirror of this code
(see the plan's §8) -- but the rule is cheap to keep and expensive to
rediscover.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, NamedTuple

from diet_guard._meal_schedule_store import current_schedule
from diet_guard._slots import day_slots

if TYPE_CHECKING:
    from collections.abc import Sequence

    from diet_guard._kuchnia_parse import Dish


class SlottedDish(NamedTuple):
    """A dish paired with the meal-slot hour it was assigned to."""

    dish: Dish
    slot: int


def assign_slots(dishes: Sequence[Dish], slots: Sequence[int]) -> list[SlottedDish]:
    """Pair each dish with a slot hour, following the provider's meal order.

    Args:
        dishes: The day's dishes, in any order (sorted here by ``priority``).
        slots: The configured slot hours, ascending, e.g. ``(8, 12, 16, 20)``.

    Returns:
        One :class:`SlottedDish` per dish, ordered by the provider's priority.
        Empty when either input is empty -- a day with no delivery and a
        schedule with no slots are both "nothing to assign", not errors.
    """
    if not dishes or not slots:
        return []
    # Ties (two dishes sharing a priority) fall back to name so the result is
    # deterministic; a bank import that reshuffles between runs would look like
    # a change and re-stamp every record.
    ordered = sorted(dishes, key=lambda dish: (dish.priority, dish.name))
    count = len(ordered)
    span = len(slots)
    return [
        SlottedDish(dish=dish, slot=slots[index * span // count])
        for index, dish in enumerate(ordered)
    ]


def dishes_in_slot_order(dishes: tuple[Dish, ...]) -> tuple[Dish, ...]:
    """Return ``dishes`` ordered by the provider's own meal priority.

    So the dish offered for the slot being filled is the one the catering
    actually intends for it, rather than whatever order the payload arrived in.
    """
    slotted = assign_slots(dishes, day_slots(current_schedule()))
    return tuple(item.dish for item in slotted)


def dish_field_values(dish: Dish) -> tuple[str, tuple[str, ...]]:
    """Return ``(portion_grams, (kcal, protein, carbs, fat))`` as form strings.

    Formatted here rather than in the widget code so the numeric formatting is
    covered by a test that needs no display.
    """
    macros = (dish.kcal, dish.protein_g, dish.carbs_g, dish.fat_g)
    return f"{dish.grams:g}", tuple(f"{value:g}" for value in macros)

File: app/lib/services/kuchnia_spread.dart
/// Assigns catering dishes to meal slots, in the provider's own order.
///
/// The Dart mirror of `diet_guard/_kuchnia_spread.py`.
///
/// The panel returns each dish with a `mealPriority` (1..N: Śniadanie, II
/// śniadanie, Obiad, Podwieczorek, Kolacja) -- the order the catering intends
/// them to be eaten in. That beats inferring an order from the dish names, and
/// it beats spreading them arithmetically.
///
/// Counts rarely match: a 5-meal plan against the default 4 slots means two
/// dishes share the first slot. The mapping is `i * S ~/ N`, which is
/// monotonic non-decreasing, keeps the first dish on the first slot and the
/// last on the last, and doubles up the earliest slots rather than dropping
/// anything.
///
/// **Integer `~/` only, never `round()`.** Repo convention for slot arithmetic
/// (`docs/meal-schedule.md`): Python's `round` is banker's and Dart's is
/// half-away-from-zero, so a float path silently desyncs the two devices --
/// and a slot one device offers while the other does not is a checkpoint that
/// can never be satisfied, i.e. a permanent lock. `index`, `span` and `count`
/// are all `int`, so `~/` matches Python's `//` exactly (`~/` on doubles
/// truncates toward zero where `//` floors).
///
/// KEEP IN SYNC WITH `diet_guard/_kuchnia_spread.py`, gated by
/// `tests/fixtures/kuchnia_day.json`.
library;

import 'package:diet_guard_app/models/kuchnia_dish.dart';

/// A dish paired with the meal-slot hour it was assigned to.
class SlottedDish {
  /// Pairs [dish] with the [slot] hour it should be logged against.
  const SlottedDish({required this.dish, required this.slot});

  /// The dish itself.
  final KuchniaDish dish;

  /// The slot hour, e.g. `12` for the 12:00 checkpoint.
  final int slot;
}

/// Pairs each dish with a slot hour, following the provider's meal order.
///
/// Returns an empty list when either input is empty -- a day with no delivery
/// and a schedule with no slots are both "nothing to assign", not errors.
List<SlottedDish> assignSlots(List<KuchniaDish> dishes, List<int> slots) {
  if (dishes.isEmpty || slots.isEmpty) return const [];

  // Ties fall back to name so the result is deterministic; a bank import that
  // reshuffled between runs would look like a change and re-stamp every
  // record.
  //
  // The payload index is the final tiebreak because `List.sort` is *not*
  // stable in Dart while Python's `sorted` is: two dishes sharing both
  // priority and name would otherwise be free to swap here and not on the PC.
  // With the index in the comparator the ordering is total, so both languages
  // agree without relying on either sort's stability.
  final indexed = <(int, KuchniaDish)>[
    for (var i = 0; i < dishes.length; i++) (i, dishes[i]),
  ];
  indexed.sort((a, b) {
    final byPriority = a.$2.priority.compareTo(b.$2.priority);
    if (byPriority != 0) return byPriority;
    final byName = a.$2.name.compareTo(b.$2.name);
    if (byName != 0) return byName;
    return a.$1.compareTo(b.$1);
  });

  final count = indexed.length;
  final span = slots.length;
  return [
    for (var index = 0; index < count; index++)
      SlottedDish(
        dish: indexed[index].$2,
        slot: slots[index * span ~/ count],
      ),
  ];
}

/// Returns [dishes] ordered by the provider's own meal priority.
///
/// So the dish offered for the slot being filled is the one the catering
/// actually intends for it, rather than whatever order the payload arrived in.
List<KuchniaDish> dishesInSlotOrder(List<KuchniaDish> dishes, List<int> slots) =>
    [for (final item in assignSlots(dishes, slots)) item.dish];
